rllib/algorithms/cql/torch/cql_torch_learner.py

```diff
@@ -0,0 +1,428 @@
+    QF_MEAN_KEY,
+    QF_MAX_KEY,
+    QF_MIN_KEY,
+    QF_PREDS,
+    TD_ERROR_MEAN_KEY,
+)
+from ray.rllib.algorithms.cql.cql import CQLConfig
+from ray.rllib.algorithms.sac.torch.sac_torch_learner import SACTorchLearner
+from ray.rllib.core.columns import Columns
+from ray.rllib.core.learner.learner import (
+    POLICY_LOSS_KEY,
+)
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.metrics import ALL_MODULES
+from ray.rllib.utils.framework import try_import_torch
+from ray.rllib.utils.typing import ModuleID, ParamDict, TensorType
+
+torch, nn = try_import_torch()
+
+
+class CQLTorchLearner(SACTorchLearner, CQLLearner):
+    @override(SACTorchLearner)
+    def compute_loss_for_module(
+        self,
+        *,
+        module_id: ModuleID,
+        config: CQLConfig,
+        batch: Dict,
+        fwd_out: Dict[str, TensorType],
+    ) -> TensorType:
+
+        # TODO (simon, sven): Add upstream information pieces into this timesteps
+        # call arg to Learner.update_...().
+        self.metrics.log_value(
+            (ALL_MODULES, TRAINING_ITERATION),
+            0
+            if math.isnan(self.metrics.peek((ALL_MODULES, TRAINING_ITERATION)))
+            else self.metrics.peek((ALL_MODULES, TRAINING_ITERATION)) + 1,
+            window=1,
+        )
+        # Get the train action distribution for the current policy and current state.
+        # This is needed for the policy (actor) loss and the `alpha`` loss.
+        action_dist_class = self.module[module_id].get_train_action_dist_cls()
+        action_dist_curr = action_dist_class.from_logits(
+            fwd_out[Columns.ACTION_DIST_INPUTS]
+        )
+
+        # Sample actions for the current state. Note that we need to apply the
+        # reparameterization trick here to avoid the expectation over actions.
+        actions_curr = (
+            action_dist_curr.rsample()
+            if not config._deterministic_loss
+            # If deterministic, we use the mean.s
+            else action_dist_curr.to_deterministic().sample()
+        )
+        # Compute the log probabilities for the current state (for the alpha loss)
+        logps_curr = action_dist_curr.logp(actions_curr)
+
+        # Optimize also the hyperparameter `alpha` by using the current policy
+        # evaluated at the current state (from offline data). Note, in contrast
+        # to the original SAC loss, here the `alpha` and actor losses are
+        # calculated first.
+        # TODO (simon): Check, why log(alpha) is used, prob. just better
+        # to optimize and monotonic function. Original equation uses alpha.
+        alpha_loss = -torch.mean(
+            self.curr_log_alpha[module_id]
+            * (logps_curr.detach() + self.target_entropy[module_id])
+        )
+
+        # Get the current batch size. Note, this size might vary in case the
+        # last batch contains less than `train_batch_size_per_learner` examples.
+        batch_size = batch[Columns.OBS].shape[0]
```

Comment:
Can we explain this logic here and why we defend against different batch sizes coming in?

Don't we expect always the same batch size from the data pipeline?

If not, we should:
a) explain here why we are expecting various batch sizes
b) probably fix this logic here. What if we call `compute_loss_for_module` 10x with a batch of `train_batch_size_per_learner - 1` (accumulating gradients for these, but not applying these) and then one batch of size `train_batch_size_per_learner`. In this case, the effective batch size would be roughly 11x the user configured one, correct?

c) Also, what if the incoming batch is larger than `train_batch_size_per_learner`?

Reply:
@sven1977 I fully agree on your comment. 

a) This was also part of the old stack algorithm. The reason behind this is that when iterating over a dataset the last batch could have less than `train_batch_size_per_learner` samples in it. 

b) The offline logic here is that after each call to `compute_loss_for_module` a `compute_gradients` and `apply_gradients` will occur. No SGD is run on the offline algorithms (yet). So the train batch size should be always as large as configured (without the last one being smaller). 

c) This case should also not happen. `iter_batches` takes care of this and ensures that always the batch size is sampled, without the last one which can be avoided by setting a flag, but neglects data.